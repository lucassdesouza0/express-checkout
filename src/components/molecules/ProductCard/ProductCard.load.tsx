import React from "react";
import Skeleton from "react-loading-skeleton";

import Button from "atoms/Button/Button";

import { Card, Details, ProductName, LowerPrice } from "./ProductCard.styles";

const ProductCard = () => {
  return (
    <Card data-testid="card" aria-labelledby="product-name">
      <Skeleton width={230} height={214} />
      <Details>
        <ProductName id="product-name">
          <Skeleton />
        </ProductName>

        <LowerPrice>
          <Skeleton />
        </LowerPrice>
      </Details>

      <Button type="button">
        <Skeleton />
      </Button>
    </Card>
  );
};

export default ProductCard;
