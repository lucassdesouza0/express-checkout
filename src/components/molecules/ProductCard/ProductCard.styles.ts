import styled from "@emotion/styled";

import mq from "utils/queries";

export const Card = styled("div")`
  display: flex;
  flex-direction: column;
  flex-shrink: 0;
  justify-content: center;
  padding: 10px;
  margin: 10px 10px;
  max-width: 250px;
  max-height: 398px;
  border-radius: 10px;
  background-color: ${(props) => (props.defaultChecked ? "#5f2eff" : "#fff")};
  color: ${(props) => (props.defaultChecked ? "#fff" : "inherit")};
  box-shadow: 2px 2px 10px 10px #eee;
  cursor: grab;

  ${mq["sm"]} {
    margin: 20px 20px;
  }
`;

export const Details = styled("div")`
  display: flex;
  flex: 1;
  flex-direction: column;
  padding: 0 10px;
`;

export const ProductName = styled("p")`
  flex: 1;
  max-width: 100%;
  font-size: 18px;
  font-weight: 400;
`;

export const LowerPrice = styled("div")`
  flex: 1;
  font-size: 18px;
  line-height: 20px;
`;
