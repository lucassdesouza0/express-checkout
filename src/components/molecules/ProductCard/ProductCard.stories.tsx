/* eslint-disable */
import React from "react";
import { storiesOf } from "@storybook/react";

import product from "../../../tests/__mocks__/product";

import ProductCard from "./ProductCard";

storiesOf("ProductCard", module).add("default", () => (
  <ProductCard product={product}></ProductCard>
));
