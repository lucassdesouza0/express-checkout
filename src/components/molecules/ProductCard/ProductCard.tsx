import React, { useContext, useState } from "react";
import Link from "next/link";
import { IProduct, IVariant } from "services/products";
import { useTranslation, Trans } from "next-i18next";

import Button from "atoms/Button/Button";
import Image from "atoms/Image/Image";
import Price from "atoms/Price/Price";
import { ContextType, ProductContext } from "pages/_app";

import { Card, Details, ProductName, LowerPrice } from "./ProductCard.styles";
import { getLowestPrice } from "utils/price/price";

interface ProductCardProps {
  product?: IProduct;
}

const ProductCard = ({ product }: ProductCardProps) => {
  const { t, ready } = useTranslation("common", { useSuspense: true });
  const { productsContext, setProductsContext } = useContext(
    ProductContext
  ) as ContextType;

  const { name, variants } = product!!;
  const [selected, setSelected] = useState(false);

  function selectProduct(product: IProduct) {
    setProductsContext({
      product: { name: product.name, variants: product.variants },
      products: productsContext.products,
    });
  }

  function lowestPrice(variants: IVariant[]) {
    const prices = variants.map((v) => v.price);

    return prices.reduce(getLowestPrice);
  }

  return (
    <Card
      data-testid="product-card"
      aria-labelledby="product-name"
      defaultChecked={selected}
    >
      <Image
        src="https://images.prismic.io/barkyn/153c4252-8084-49b8-9931-3ec426cdd859_premium_box.png?auto=compress,format"
        width="230"
        height="214"
      />
      <Details>
        <ProductName id="product-name">
          <Trans i18nKey="products">{name}</Trans>
        </ProductName>
        <LowerPrice>
          <label htmlFor="">{t("from")}</label>
          <Price value={lowestPrice(variants)} />
        </LowerPrice>
      </Details>

      <Button type="button" onClick={() => selectProduct(product!!)}>
        <Link href="#pre-checkout">{t("select")}</Link>
      </Button>
    </Card>
  );
};

export default ProductCard;
