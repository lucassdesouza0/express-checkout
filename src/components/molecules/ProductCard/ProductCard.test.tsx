import React from "react";
import { screen, render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import testRender from "../../../tests/render";

import mock from "../../../tests/__mocks__/product";

import ProductCard from "./ProductCard";

describe("<ProductCard />", () => {
  test("it should mount", () => {
    testRender(<ProductCard product={mock}></ProductCard>);

    const card = screen.getByTestId("product-card");

    expect(card).toBeInTheDocument();
  });

  test("snapshot", () => {
    const { asFragment } = render(<ProductCard product={mock}></ProductCard>);
    expect(asFragment()).toMatchSnapshot();
  });
});
