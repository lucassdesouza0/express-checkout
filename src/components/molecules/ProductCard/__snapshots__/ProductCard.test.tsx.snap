// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCard /> it should match snapshot 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="product-name"
    class="css-zmijso-Card ek8h0p83"
    data-testid="product-card"
  >
    <img
      class="css-1ep1cpx-SImage e65m3ty0"
      data-testid="image"
      height="214"
      src="https://images.prismic.io/barkyn/153c4252-8084-49b8-9931-3ec426cdd859_premium_box.png?auto=compress,format"
      width="230"
    />
    <div
      class="css-yf3tg6-Details ek8h0p82"
    >
      <p
        class="css-hbd1xf-ProductName ek8h0p81"
        id="product-name"
      >
        Dry Dog Food - Chicken
      </p>
      <div
        class="css-1wmjec5-LowerPrice ek8h0p80"
      >
        <label
          for=""
        >
          from
        </label>
      </div>
    </div>
    <button
      class="css-hhtkyx-Button enhitpr0"
      name="button"
      type="button"
    >
      <a
        href="/#pre-checkout"
      >
        select
      </a>
    </button>
  </div>
</DocumentFragment>
`;

exports[`<ProductCard /> it should render the load version of the card 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="product-name"
    class="css-zmijso-Card ek8h0p83"
    data-testid="card"
  >
    <span>
      <span
        class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
        style="width: 230px; height: 214px;"
      >
        ‌
      </span>
    </span>
    <div
      class="css-yf3tg6-Details ek8h0p82"
    >
      <p
        class="css-hbd1xf-ProductName ek8h0p81"
        id="product-name"
      >
        <span>
          <span
            class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
          >
            ‌
          </span>
        </span>
      </p>
      <div
        class="css-1wmjec5-LowerPrice ek8h0p80"
      >
        <span>
          <span
            class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
          >
            ‌
          </span>
        </span>
      </div>
    </div>
    <button
      class="css-hhtkyx-Button enhitpr0"
      name="button"
      type="button"
    >
      <span>
        <span
          class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
        >
          ‌
        </span>
      </span>
    </button>
  </div>
</DocumentFragment>
`;
